### PR TITLE
Sbml factory

### PR DESCRIPTION
**Pull request by cshenry**
Completed implementation of SBML import. Example command:
ms import sbml iJR904.xml iJR904.sbml

You can obtain SBML file here:
http://bioseed.mcs.anl.gov/~chenry/ModelSEED/iJR904.xml

This creates a new model, biochemistry, annotation, and mapping in your database, all called "iJR904.sbml".

### CODE REVIEW

File: lib/ModelSEED/App/import/Command/sbml.pm [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+sub execute {
+    my ($self, $opts, $args) = @_;
+    my $auth  = ModelSEED::Auth::Factory->new->from_config;
+    my $store = ModelSEED::Store->new(auth => $auth);
+    my $helper = ModelSEED::App::Helpers->new();
+    my $factory = ModelSEED::MS::Factories::SBMLFactory->new({
+    	auth => $auth,
+    	store => $store
+    });
+    print($self->usage) && exit if $opts->{help};
```
Comment (devoid) [inline lib/ModelSEED/App/import/Command/sbml.pm]: Replace exit with return.

File: lib/ModelSEED/MS/Factories/SBMLFactory.pm [hunk 2]
```diff
@@ -0,0 +1,670 @@
...
+
+The filename of the SBML file to be imported. Required.
+
+=item annotation
+
+A L<ModelSEED::MS::Annotation> object. Required.
+
+=back
+
+=cut
```
Comment (devoid) [inline lib/ModelSEED/MS/Factories/SBMLFactory.pm:44]: POD is out of date: rename function to parseSBML, returns a list of objects, etc.
Also include expected usage for factory constructor (new).

File: lib/ModelSEED/MS/Factories/SBMLFactory.pm [hunk 3]
```diff
@@ -0,0 +1,670 @@
...
+    ModelSEED::MS::Annotation
+    ModelSEED::MS::Mapping
+);
+use Try::Tiny;
+use Moose;
+use namespace::autoclean;
+
+has auth => ( is => 'ro', isa => "ModelSEED::Auth", required => 1);
+has store => ( is => 'ro', isa => "ModelSEED::Store", required => 1);
+has auth_config => ( is => 'ro', isa => 'HashRef', lazy => 1, builder => '_build_auth_config');
```
Comment (devoid) [inline lib/ModelSEED/MS/Factories/SBMLFactory.pm]: These don't seem to be used. Can they be removed?

Comment (cshenry) [inline lib/ModelSEED/MS/Factories/SBMLFactory.pm]: Good comments. Should I implement the changes?

File: lib/ModelSEED/MS/Factories/SBMLFactory.pm [hunk 4]
```diff
@@ -0,0 +1,670 @@
...
+		});
+		$args->{annotation}->add("genomes",{
+			id => $mdlData->{id},
+			name => $mdlData->{name},
+			source => "SBML"
+		});
+		$newAnno = $args->{annotation};
+    } elsif (!defined($args->{mapping})) {
+    	$args->{mapping} = $args->{annotation}->mapping();
+    }
```
Comment (devoid) [inline lib/ModelSEED/MS/Factories/SBMLFactory.pm]: Not sure I understand what's happening here with Biochemistry, Annotation and Mapping.
1. I think you always construct a biochemistry, but if you already have an associated annotation / mapping, how does the new bio get linked? Especially what happens if it doesn't contain the same reactions?
1. What happens if I provide an annotation but not a mapping? A mapping but not an annotation? Both? Niether?

File: lib/ModelSEED/MS/Factories/SBMLFactory.pm [hunk 5]
```diff
@@ -0,0 +1,670 @@
...
+	} elsif ($root =~ m/\+/) {
+		$proteins->[0] = $self->parseSingleProtein($root,$gprHash);
+	} elsif (defined($gprHash->{$root})) {
+		$gprHash->{root} = $gprHash->{$root};
+		return $self->translateGPRHash($gprHash);
+	} else {
+		$proteins->[0]->[0]->[0] = $root;
+	}
+	return $proteins;
+}
```
Comment (devoid) [inline lib/ModelSEED/MS/Factories/SBMLFactory.pm:664]: What does `$gprHash` look like?
What should `$proteins` look like? Especially that triple-deep array ref...
At each if statement, what do we find out about the gprHash?

File: Build.PL [hunk 6]
```diff
@@ -108,6 +108,7 @@ my $builder = $subclass->new(
         "YAML" => 0,
         "common::sense" => 0,
         "namespace::autoclean" => 0,
+        "XML::DOM::Parser" => 0,
```
Comment (devoid) [inline Build.PL]: The module is XML::DOM. Parser is part of that.

File: lib/ModelSEED/App/import/Command/sbml.pm [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+use base 'App::Cmd::Command';
+use Class::Autouse qw(
+    ModelSEED::Store
+    ModelSEED::Auth::Factory
+    ModelSEED::MS::Factories::SBMLFactory
+    ModelSEED::Database::Composite
+    ModelSEED::Reference
+    ModelSEED::App::Helpers
+);
+
```
Comment (devoid) [inline lib/ModelSEED/App/import/Command/sbml.pm:14]: Add ModelSEED::utilities qw( args verbose set_verbose );

**Comment by devoid** [comment]
First off, this is awesome. I only checked the iJR904 model, but it worked, producing it's own annotation, mapping, biochemistry and model! There's a lot of work going on in here.

In addition to the line-by-line comments I put down, I'd like to see better comments on the helper functions. In particular, `parseSingleProtein`, `translateGPRHash` and `parseGPR` could use a description of the input and output arguments, and a few single-line comments explaining what's going on in the major blocks.

File: lib/ModelSEED/MS/Factories/SBMLFactory.pm (hunk 4 above)
Comment (cshenry) [inline lib/ModelSEED/MS/Factories/SBMLFactory.pm]: Well, if there is a piece of this that is iffy, it's this.
The safest thing to do is probably to not allow the user to provide any annotation or mapping and always build a new one.
We can merge in an annotation object in a separate function.
I'll make that change.

File: lib/ModelSEED/MS/Factories/SBMLFactory.pm (hunk 5 above)
Comment (cshenry) [inline lib/ModelSEED/MS/Factories/SBMLFactory.pm:664]: Well, it works, and it complains if it doesn't work. I'm not going to re-implement that particular thing.
Obviously, it is a complex piece of code that can take any GPR string and convert to our protein data structure. This does it, and it works when applied to an entire genome-scale model. And it works right. Not sure what else I have to say here.
Chris

Comment (devoid) [inline lib/ModelSEED/MS/Factories/SBMLFactory.pm:664]: If you could just add comments in to explain what it's doing, where info is coming from and how it's being manipulated that would be great. No rewrite necessary. 

~ Scott

**Comment by mdejongh** [comment]
I tried this out and it worked, but the reactions print with long
compartment names (like [External_compartment] ...) - should you map these
to individual characters?

On Mon, Oct 22, 2012 at 6:24 PM, cshenry notifications@github.com wrote:

> Completed implementation of SBML import. Example command:
> ms import sbml iJR904.xml iJR904.sbml
> 
> You can obtain SBML file here:
> http://bioseed.mcs.anl.gov/~chenry/ModelSEED/iJR904.xml
> 
> This creates a new model, biochemistry, annotation, and mapping in your
> 
> ## database, all called "iJR904.sbml".
> 
> You can merge this Pull Request by running:
> 
>   git pull https://github.com/ModelSEED/ModelSEED SBMLFactory
> 
> Or view, comment on, or merge it at:
> 
>   https://github.com/ModelSEED/ModelSEED/pull/129
> Commit Summary
> - Adding commands to support import of SBML files
> - Merge branch 'master' into SBMLFactory
> - Minor fixes to SBML command
> - Completing implementation of SBML import
> 
> File Changes
> - _M_ Build.PL (1)
> - _A_ lib/ModelSEED/App/import/Command/sbml.pm (88)
> - _A_ lib/ModelSEED/MS/Factories/SBMLFactory.pm (670)
> 
> Patch Links
> - https://github.com/ModelSEED/ModelSEED/pull/129.patch
> - https://github.com/ModelSEED/ModelSEED/pull/129.diff
>   
>   —
>   Reply to this email directly or view it on GitHubhttps://github.com/ModelSEED/ModelSEED/pull/129.

## 

Matt DeJongh, Ph.D.
Computer Science Department
Hope College, Holland, MI 49423
(616) 395-7429

**Comment by cshenry** [comment]
Yeah, I considered this. I'm a bit torn. Part of me says, keep the models completely unaltered. However, I do think we should make some effort to standardize compartment names and provide compartment abbreviations. I'll make those changes. Scott had a number of good recommendations as well.

**Comment by cshenry** [comment]
OK. All requested changes made. This should be ready to merge.

**Pull request merged**